playlistmanager.lua

```diff
@@ -275,11 +275,32 @@ elseif settings.default_sort == 'size-desc' then
   sort_mode = 5
 end
 
+function is_protocol(path)
+  return type(path) == 'string' and path:match('^%a[%a%d-_]+://') ~= nil
+end
+
+--escape special characters in url
+function url_decode(url)
+  local function hex_to_char(x)
+    return string.char(tonumber(x, 16))
+  end
+  if url ~= nil then
+    url = url:gsub('^file://', '')
+    url = url:gsub('%%(%x%x)', hex_to_char)
+    if url:match('://localhost:?') then
+      url = url:gsub('^.*/', '')
+    end
```

Comment:
I think this could use the filename_replace. The only thing it cannot do is the hex_to_char conversion, since the rules are serialized as a string to support conf file.

I think we could solve it like this. Add a custom magic string to represent utility functions like:
```lua
filename_replace_functions = {
  hex_to_char=function(x) return string.char(tonumber(x, 16)) end
}
...
-- line 351 file name replacing
         for rule, override in pairs(indexrules) do
            override = filename_replace_functions[override] or override
            tmp = tmp:gsub(rule, override)
          end
```

Currently filename replacing is disabled by default, but maybe we could implement some reasonable defaults. What do you think?

Reply:
This seems reasonable, and it is much better to implement it in this `stripfilename` function. I will try it.

Reply:
Try the following patch, but it doesn't work.
```diff
Subject: [PATCH] escape special characters in url

---
 playlistmanager.lua | 16 +++++++++++++++-
 1 file changed, 15 insertions(+), 1 deletion(-)

diff --git a/playlistmanager.lua b/playlistmanager.lua
index 3427151..9d2510c 100644
--- a/playlistmanager.lua
+++ b/playlistmanager.lua
@@ -38,7 +38,16 @@ local settings = {
   --'all' will match any extension or protocol if it has one
   --uses json and parses it into a lua table to be able to support .conf file
 
-  filename_replace = "",
+  filename_replace = [[
+    [
+      {
+        "protocol": { "all": true },
+        "rules": [
+          { "%%(%x%x)": "hex_to_char" }
+        ]
+      }
+    ]
+  ]],
 
 --[=====[ START OF SAMPLE REPLACE - Remove this line to use it
   --Sample replace: replaces underscore to space on all files
@@ -354,12 +363,17 @@ function stripfilename(pathfile, media_title)
   local protocol = is_protocol(pathfile)
   if not ext then ext = "" end
   local tmp = pathfile
+  --escape special characters in url
+  filename_replace_functions = {
+    hex_to_char = function(x) return string.char(tonumber(x, 16)) end
+  }
   if settings.filename_replace and not media_title then
     for k,v in ipairs(settings.filename_replace) do
       if ( v['ext'] and (v['ext'][ext] or (ext and not protocol and v['ext']['all'])) )
       or ( v['protocol'] and (v['protocol'][protocol] or (protocol and not ext and v['protocol']['all'])) ) then
         for ruleindex, indexrules in ipairs(v['rules']) do
           for rule, override in pairs(indexrules) do
+            override = filename_replace_functions[override] or override
             tmp = tmp:gsub(rule, override)
           end
         end
-- 
```

Reply:
```
local protocol = pathfile:match("^(%a%a+)://")
```
Is this returning a value for your file?

Reply:
Maybe issue is actually in `v['protocol'] and (v['protocol'][protocol] or (protocol and not ext and v['protocol']['all']))`

Maybe needs some simplification.

Reply:
> ```
> local protocol = pathfile:match("^(%a%a+)://")
> ```
> 
> Is this returning a value for your file?

![image](https://user-images.githubusercontent.com/61936050/206849751-f3054654-a04f-41f0-b299-1255fe7d2134.png)
It is.

Reply:
> Maybe issue is actually in `v['protocol'] and (v['protocol'][protocol] or (protocol and not ext and v['protocol']['all']))`
> 
> Maybe needs some simplification.

yes. The problem is `not ext`, It is unnecessary.

Reply:
```diff
diff --git a/playlistmanager.lua b/playlistmanager.lua
index 653a0de..1dde52a 100644
--- a/playlistmanager.lua
+++ b/playlistmanager.lua
@@ -38,7 +38,16 @@ local settings = {
   --'all' will match any extension or protocol if it has one
   --uses json and parses it into a lua table to be able to support .conf file
 
-  filename_replace = "",
+  filename_replace = [[
+    [
+      {
+        "protocol": { "all": true },
+        "rules": [
+          { "%%(%x%x)": "hex_to_char" }
+        ]
+      }
+    ]
+  ]],
 
 --[=====[ START OF SAMPLE REPLACE - Remove this line to use it
   --Sample replace: replaces underscore to space on all files
@@ -126,7 +135,7 @@ local settings = {
   --####  VISUAL SETTINGS
 
   --prefer to display titles for following files: "all", "url", "none". Sorting still uses filename.
-  prefer_titles = "url",
+  prefer_titles = "none",
 
   --call youtube-dl to resolve the titles of urls in the playlist
   resolve_titles = false,
@@ -345,6 +354,20 @@ function escapepath(dir, escapechar)
   return string.gsub(dir, escapechar, '\\'..escapechar)
 end
 
+function replace_lefthand_matches(value, valid_values)
+  if protocol == nil or valid_values == nil then
+    return false
+  end
+  if valid_values['all'] then
+    return true
+  end
+  return valid_values['all'] or valid_values[protocol]
+end
+
+local filename_replace_functions = {
+  hex_to_char = function(x) return string.char(tonumber(x, 16)) end
+}
+
 --strip a filename based on its extension or protocol according to rules in settings
 function stripfilename(pathfile, media_title)
   if pathfile == nil then return '' end
@@ -354,10 +377,10 @@ function stripfilename(pathfile, media_title)
   local tmp = pathfile
   if settings.filename_replace and not media_title then
     for k,v in ipairs(settings.filename_replace) do
-      if ( v['ext'] and (v['ext'][ext] or (ext and not protocol and v['ext']['all'])) )
-      or ( v['protocol'] and (v['protocol'][protocol] or (protocol and not ext and v['protocol']['all'])) ) then
+      if replace_lefthand_matches(ext, v['ext']) or replace_lefthand_matches(protocol, v['protocol']) then
         for ruleindex, indexrules in ipairs(v['rules']) do
           for rule, override in pairs(indexrules) do
+            override = filename_replace_functions[override] or override
             tmp = tmp:gsub(rule, override)
           end
         end
```

Reply:
Done.